txpipe/magnification.py

```diff
@@ -0,0 +1,204 @@
+            w2 = mag_cat[f'lens/bin_{ibin}/weight'][:]
+
+            #compute mag coeff with the whole sample
+            csample = self.calc_frac_change(w1, w2)/deltak
+            self.write_output(outfile, csample, ibin)
+
+            if self.config["bootstrap_error"]:
+                print(f'Computing uncertainty with bootstrap')
+
+                #compute mag coeff with subsampling patches to get a boostrap error
+                patch1 = cluster.predict(
+                    np.transpose(
+                        [ nomag_cat[f'lens/bin_{ibin}/ra'][:],
+                          nomag_cat[f'lens/bin_{ibin}/dec'][:], ]
+                        )
+                    )
+                patch2 = cluster.predict(
+                    np.transpose(
+                        [ mag_cat[f'lens/bin_{ibin}/ra'][:],
+                          mag_cat[f'lens/bin_{ibin}/dec'][:], ]
+                        )
+                    )
+
+                #count number of objects in each patch
+                label1, counts1 = np.unique(patch1, return_counts=True)
+                label2, counts2 = np.unique(patch2, return_counts=True)
+                assert (label1 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+                assert (label2 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+
+                # define a function that computes fractional number count change
+                # and takes only patch IDs as input (i.e. fix the counts to this z bin)
+                def calc_frac_change_patches_ibin(patch_ids):
+                    return self.calc_frac_change_patches(patch_ids, counts1, counts2)
+
+                boot = bootstrap(np.array([label1]), calc_frac_change_patches_ibin)
+
+                csample_boot_mean = np.mean(boot.bootstrap_distribution)/deltak
+                csample_boot_err = np.std(boot.bootstrap_distribution)/np.abs(deltak)
+                self.write_output_boot(outfile, csample_boot_mean, csample_boot_err, ibin)
+
+
+        self.plot_results(outfile)
+        outfile.close()
+
+    def calc_cluster_patches(self, nomag_cat ):
+        """
+        Split the lens sample into patches 
+        We are using the full no-magnification sample to do this
+        but we could also use jackknife patches 
+        based on this SSI run's mask if we have it...
+        """
+        import sklearn.cluster
+
+        n_patches = self.config["n_patches"]
+        cluster = sklearn.cluster.KMeans(n_clusters=n_patches, n_init='auto')
+
+        # limit number of objects too 1000 per patch so stop the training 
+        # from taking too long  
+        max_per_patch = 1000
+        maxobj = max_per_patch*n_patches
+
+        n_obj_lens = nomag_cat[f'lens/bin_all/ra'].shape[0]
+
+        if n_obj_lens > maxobj:
+            index = np.sort(np.random.choice(n_obj_lens, size=maxobj, replace=False))
+            ra = nomag_cat[f'lens/bin_all/ra'][index]
```

Comment:
Random access reads like this using HDF can be painfully slow.  It's definitely worth sorting the index before using it, but also leaving a note that if it's too slow then we can look at ways of speeding it up.

Reply:
done